### PR TITLE
ARROW-16439: [R] Implement binding for `lubridate::fast_strptime`

### PR DESCRIPTION
**Pull request by dragosmg**
This PR adds:
* a binding emulating lubridate's `fast_strptime` functionality  

This PR does not support the following arguments:
* `lt = TRUE` - this returns a POSIXlt object (a list) which cannot be easily used in a dplyr pipeline (currently it actually errors) => we have a different default `lt = FALSE` for the Arrow binding, and
* `cutoff_2000 = 68L` - for the `y%` format two-digit numbers smaller or equal to `cutoff_2000` are parsed as though starting with 20, otherwise parsed as though starting with 19. It would be nice to have this, so I raised [ARROW-16596](https://issues.apache.org/jira/browse/ARROW-16596). We can always suggest users they manipulate the strings before parsing, so I don't think this is crucial functionality. 

The following code will be possible once the PR is merged:
``` r
library(lubridate, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
library(arrow, warn.conflicts = FALSE)

dates_table <- tibble(
  string_with_short_year = c("68-05-17", "69-05-17", "55-05-17")
)

dates_table %>% 
  mutate(
    date = fast_strptime(
      string_with_short_year, 
      format = "%y-%m-%d",
      lt = FALSE
    )
  )
#> # A tibble: 3 × 2
#>   string_with_short_year date               
#>   <chr>                  <dttm>             
#> 1 68-05-17               2068-05-17 00:00:00
#> 2 69-05-17               1969-05-17 00:00:00
#> 3 55-05-17               2055-05-17 00:00:00

dates_table %>% 
  arrow_table() %>% 
  mutate(
    date = fast_strptime(
      string_with_short_year, 
      format = "%y-%m-%d",
      lt = FALSE
    )
  ) %>%
  collect()
#> # A tibble: 3 × 2
#>   string_with_short_year date               
#>   <chr>                  <dttm>             
#> 1 68-05-17               2068-05-17 00:00:00
#> 2 69-05-17               1969-05-17 00:00:00
#> 3 55-05-17               2055-05-17 00:00:00
```

<sup>Created on 2022-05-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: r/R/dplyr-funcs-datetime.R [hunk 1]
```diff
@@ -618,4 +618,40 @@ register_bindings_datetime_parsers <- function() {
...
+  register_binding("fast_strptime", function(x,
+                                             format,
+                                             tz = "UTC",
+                                             lt = FALSE,
+                                             cutoff_2000 = 68L) {
+    # `lt` controls the output `lt = TRUE` returns a POSIXlt (which doesn't play
+    # well with mutate, for example)
+    if (lt) {
+      arrow_not_supported("`lt = TRUE` argument")
+    }
```
Comment (thisisnic) [inline r/R/dplyr-funcs-datetime.R:631]: Please can you chuck in a test for `lt=TRUE`?

File: r/R/dplyr-funcs-datetime.R [hunk 2]
```diff
@@ -618,4 +618,40 @@ register_bindings_datetime_parsers <- function() {
...
+
+    # TODO revisit once https://issues.apache.org/jira/browse/ARROW-16596 is done
+    if (cutoff_2000 != 68L) {
+      arrow_not_supported("`cutoff_2000` != 68L argument")
+    }
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:636]: I think that is tested here:

https://github.com/dragosmg/arrow/blob/64230541b6cf5ebe3131a419b7dfb8a85fe8d5c8/r/tests/testthat/test-dplyr-funcs-datetime.R#L1891-L1909

File: r/tests/testthat/test-dplyr-funcs-datetime.R [hunk 3]
```diff
@@ -1812,3 +1812,98 @@ test_that("ym, my & yq parsers", {
...
+            format = "%y-%m-%d %H:%M:%S",
+            lt = FALSE,
+            cutoff_2000 = 69L
+          )
+      ) %>%
+      collect(),
+    tibble(
+      x = c("68-10-07 19:04:05", "69-10-07 19:04:05", NA)
+    ),
+    warning = TRUE
```
Comment (thisisnic) [inline r/tests/testthat/test-dplyr-funcs-datetime.R:1907]: Ah wait, I see you've added the test for a warning here, but I think `expect_warning()` or `expect_error()` would be simpler?

File: r/R/dplyr-funcs-datetime.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/R/dplyr-funcs-datetime.R:636]: Never mind - I see you've done this below.

File: r/tests/testthat/test-dplyr-funcs-datetime.R (hunk 3 above)
Comment (thisisnic) [reply r/tests/testthat/test-dplyr-funcs-datetime.R:1907]: Hmm, we appear to have a mix of both, but given that the `warning` parameter for `compare_dplyr_binding()` appears to be for this purpose, then that's all good.

**Review by thisisnic — APPROVED** [review]
Looks good to me!

File: r/R/dplyr-funcs-datetime.R [hunk 4]
```diff
@@ -618,4 +618,40 @@ register_bindings_datetime_parsers <- function() {
   for (ymd_order in ymd_parser_vec) {
     register_binding(ymd_order, ymd_parser_map_factory(ymd_order))
   }
+
+  register_binding("fast_strptime", function(x,
+                                             format,
+                                             tz = "UTC",
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:624]: Could we add a test for non-UTC timezones as well? I don't see any below, I don't think?

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:624]: Good catch. Thanks. Done!

**Review by jonkeane — COMMENTED** [review]
One last cleanup

File: r/tests/testthat/test-dplyr-funcs-datetime.R [hunk 5]
```diff
@@ -1812,3 +1812,133 @@ test_that("ym, my & yq parsers", {
...
+            lt = FALSE
+          )
+      ) %>%
+      collect(),
+    tibble(
+      x =
+        c("68-10-07 19:04:05", "69-10-07 19:04:05", NA)
+    )#,
+    # arrow does not preserve the `tzone` attribute
+    # test ignore_attr = TRUE
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-datetime.R:1905]: I yes, sorry. Forgot about those. Do I open a minor PR?

[the remark this answers is not in the document]
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-datetime.R:1905]: Or should I just do it in one of the other PR I have going?

[the remark this answers is not in the document]
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-datetime.R:1905]: Removed https://github.com/dragosmg/arrow/commit/bde85ce3d7133094ac067d7092f9c016e1f9f72f

**Pull request closed without merging**